### ISSUE TITLE
Extend RollExpressions to parse skill expressions

### ISSUE THREAD

**Issue by trey-kirk-sp** [mr-ice/maptool-macros#158]
So with #152  much of the token's calculations have been stuffed into properties. This way changes to the base ability scores are immediately reflected in the related skill and save checks. However, attacks and spells are still rather static due to their close relationship to RollExpressions. So to cross that gap, we need to allow expressions that expand property values in RollExpressions, ex, when defining an attack expression, set the bonus to ${bonus.attack.melee}. Spell attack bonus would have to be a little more verbose due to spellcasting ability: ${Proficiency} + ${WisdomBonus}. Brings up another question: are arithmetic expressions supported? We'll have to make sure they are.

- [x] Write an expression parser to expand property values and resolve math expressions.
- [x] Hook parser into dnd5e_RollExpression_getBonus. Consider if dnd5e_RollExpression_getRollString needs it, also.
- [x] Delete SpellcastingAbility property from campaign properties. Doesn't really make sense at the token level for multi-classing spellcaster.
- [x] Refactor Spells and Attacks DnDBeyond DTOs to include property expressions instead of literal values.
- [x] Test AE2 to see if there's any problems using property expressions.

**Comment by trey-kirk-sp** [comment]
Consider taking advantage macroEval to solve both parsing requirements. So instead of inventing a property expression, just parse the value as a macro or property string. This route opens the risk of allowing the user to supply arbitrary macro code that could be executed by a trusted macro. So that could be awesome.

**Comment by trey-kirk-sp** [comment]
Holy shit this was way easier than it should've been. Just needed to update dnd5e_RollExpression_getBonus like so:
```
[h: bonus = json.get (arg (0), "bonus")]
[h: propertyPattern = "\\{.*\\}"]
<!-- If the bonus has a property string in it and we have a current token, expand the value -->
[h, if (currentToken() != "" && matches (bonus, propertyPattern)): bonus = evalMacro (bonus); ""]
[h, if (!isNumber(bonus)): bonus = 0; ""]
[h: macro.return = bonus]
```
Works weirdly like a charm. I'm just not used to immediate success from MT macros like this.

I need a minute...

**Comment by trey-kirk-sp** [comment]
Oh, whew! This isn't as straight-forward as MT lead me to believe. Properties like {Animal Handling} will resolve in a way that's very sad and heartbreaking. Also, the arithmetic starts to fall off when you do stuff like  1d20 - {Proficiency} - {CharismaBonus}. I think I'll stick with the existing notation because it feels natural to wrap properties in { }. But I'll parse out the properties myself via regex and expand them via getProperty(). The last step is just forcing a string of literal values like, "2 + 7 - 8" to math itself out.

**Comment by trey-kirk-sp** [comment]
So this is the halfway point: Types have been prototyped and now define what Rollers apply to them. This allows us to add new Types as needed, and from other token libraries, without refactoring dnd5e Dice Roller. I did that so I could create a Weapon and Proficient Type. This way our editors can easily mark weapon attacks and denote proficiency without getting mired in parsing property expressions. At the same time, property expressions can still be added adhoc, if somebody really wanted to.

The next step is to go through AE2 and get it to at least get away from setting string Types. I'll probably have to work with Jay to figure out how to update the UI to add some new fields.

**Comment by trey-kirk-sp** [comment]
Todo: simplify rerolls by having Rollers just use counters instead of modding the types list

**Comment by trey-kirk-sp** [comment]
Almost there...

- [ ] AE1 isn't parsing extra damage
- [ ] Tested the crap out of weapon attack rolls. Need to build some tests around Save Damage/Save Effect
- [ ] Take another pass at performance improvements

**Comment by trey-kirk-sp** [comment]
Ah nuts. All of the nifty Save Damage processing Jay wrote breaks with the new Dice Roller stack model, because I moved all of the totalling logic out of RE_getTotal to save some cycles. Easy to fix, just make a DiceRoller_totaler with priority 100. Rollers that need to modify or know the total roll (saveDamageRoll, saveEffectRoll) should be given a later priority. Leave DR_finalizer to build the descriptors.

**Comment by trey-kirk-sp** [comment]
Fixed in PR #163
